feat: auto format `.mdx` files with husky and prettier

gulshanpr opened the pull request:
<!--
Please read and fill out this form before submitting your PR.

Please make sure you have reviewed our contributors guide before submitting your
first PR.
-->

## Overview
issue #676
prev. PR #1010 [closed by mistake, apology for that]

added the suggested changes by @rootulp 
https://github.com/celestiaorg/docs/pull/1010#discussion_r1317721546

this PR does
1. auto format the `.mdx` file of `docs` with prettier, when committing (with Husky)
2. installed prettier in `github workflow` that fails if not formatted 


<!-- 
Please provide an explanation of the PR, including the appropriate context,
background, goal, and rationale. If there is an issue with this information,
please provide a tl;dr and link the issue. 
-->

## Checklist

<!-- 
Please complete the checklist to ensure that the PR is ready to be reviewed.

IMPORTANT:
PRs should be left in Draft until the below checklist is completed.
-->

- [x] New and updated code has appropriate documentation
- [x] New and updated code has new and/or updated testing
- [x] Required CI checks are passing
- [x] Visual proof for any user facing features like CLI or documentation updates
- [x] Linked issues closed with keywords

jcstein commented:
it looks like the linting worked! should it have modified the files with husky as well? avoiding the lint error on build? or would this only happen once this PR is merged?

jcstein commented:
For some reason in the past, I've ran into issues when using this syntax for mdx-code-blocks

<img width="151" alt="screenshot 2023-09-08 at 12 54 51" src="https://github.com/celestiaorg/docs/assets/46639943/bd665aaa-d727-4a3d-9e83-4e9b606747e5">

and have only had luck with four ticks

<img width="168" alt="screenshot 2023-09-08 at 12 55 12" src="https://github.com/celestiaorg/docs/assets/46639943/b919f25c-dfef-49cf-bd76-a49693fc52d7">

can you confirm the failing linting is not due to the instances of four backticks?

gulshanpr commented:
> thank you for the PR!
> 
> it looks like the linting worked! should it have modified the files with husky as well? avoiding the lint error on build? or would this only happen once this PR is merged?

i made few changes, previously it was only modifying and not committing with the current commit, but now with this
 `npx lint-staged && git add . && git commit --no-verify -m 'Apply linting and formatting'`
it will modify and commit

gulshanpr commented:
> 
> For some reason in the past, I've ran into issues when using this syntax for mdx-code-blocks
> <img alt="screenshot 2023-09-08 at 12 54 51" width="151" src="https://user-images.githubusercontent.com/46639943/266680408-bd665aaa-d727-4a3d-9e83-4e9b606747e5.png">
> 
> and have only had luck with four ticks
> <img alt="screenshot 2023-09-08 at 12 55 12" width="168" src="https://user-images.githubusercontent.com/46639943/266680484-b919f25c-dfef-49cf-bd76-a49693fc52d7.png">
> 
> can you confirm the failing linting is not due to the instances of four backticks?
 
no this is not happening either with ```(three) or ````(four) backticks, prettier is fixing minor issue that is causing the linting error.

jcstein commented on .github/workflows/lint.yml:
```diff
@@ -17,4 +17,9 @@ jobs:
+      - name: Install Prettier
+        run: npm install -g prettier
+
+      - name: lint .mdx file
```
```suggestion
      - name: Lint .mdx files
```

jcstein commented:
fantastic contribution @gulshanpr - would you like to add a little bit in the readme on how to use prettier to format and how to check with the linter before committing? also maybe run .md files through husky as well?

gulshanpr commented:
> fantastic contribution @gulshanpr - would you like to add a little bit in the readme on how to use prettier to format and how to check with the linter before committing? also maybe run .md files through husky as well?

Yeah, I can make a PR

jcstein commented:
can you add to this instead?

gulshanpr commented:
[problem] `only` problem I ran into is, prettier is changing `*` into `-` this for sub-points. I tried couple of way and it doesn't worked.
updated README for linting and formatting with prettier before committing.
also implemented linting for `.md` files. It auto run after committing

jcstein commented:
> [problem] only problem I ran into is, prettier is changing * into - this for sub-points. I tried couple of way and it doesn't worked.

this is interesting, as I think it will then cause issues with linting on the CI here, causing it to fail, right? maybe you could write a special rule for this case?

gulshanpr commented:
> this is interesting, as I think it will then cause issues with linting on the CI here, causing it to fail, right? maybe you could write a special rule for this case?

1. no, it will not. I have checked the lint script locally and it ran successfully for both `.md` and `.mdx` files, so it will not fail in CI check.
2. the changes that made by prettier (by auto formatting many files) , i checked they all are following the format standard of `prettier` and `markdown`. this will always make sure to maintain the formatting rules of prettier because formatting will auto trigger on committing any changes. 
 👍

jcstein commented:
awesome, I just think conflicts will need to be resolved and then this LGTM!

jcstein commented:
"Deploy PR previews" is expected to fail

For the linting, I resolved some conflicts but didn't format properly. Now, I'm trying to use the formatting you setup following the readme:

```bash
jcs$yarn lint-check
yarn run v1.22.19
$ prettier --check docs/**/*.md --ignore-unknown && prettier --check docs/**/*.mdx --ignore-unknown
Checking formatting...
[warn] docs/developers/overview.md
[warn] Code style issues found in the above file. Run Prettier to fix.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
jcs$npx lint-staged
→ No staged files found.
```

What step am I missing for `npx lint-staged` to format the files properly?

jcstein commented:
and should Husky prevent me from committing if the linting doesn't pass?

gulshanpr commented:
> "Deploy PR previews" is expected to fail
> 
> For the linting, I resolved some conflicts but didn't format properly. Now, I'm trying to use the formatting you setup following the readme:
> 
> ```shell
> jcs$yarn lint-check
> yarn run v1.22.19
> $ prettier --check docs/**/*.md --ignore-unknown && prettier --check docs/**/*.mdx --ignore-unknown
> Checking formatting...
> [warn] docs/developers/overview.md
> [warn] Code style issues found in the above file. Run Prettier to fix.
> error Command failed with exit code 1.
> info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
> jcs$npx lint-staged
> → No staged files found.
> ```
> 
> What step am I missing for `npx lint-staged` to format the files properly?

to format the code automatically with `npx lint-staged` command, few conditions are need to be met,
1. `npx lint-staged` only work for the files that are in staging area (not work for the file that are not formatted and not in staging are)

you need to add the unfomatted file first(`git add`) then, `npx lint-staged` command it will work then,

gulshanpr commented:
> and should Husky prevent me from committing if the linting doesn't pass?

yes, with husky you can not commit until all files are formatted

jcstein commented:
> you need to add the unfomatted file first(git add) then, npx lint-staged command it will work then,
(i didn't specify this in README)

do you mind specifying in readme? i'll cleanup the conflicts now.

jcstein commented:
this is getting really close. i was testing the `npx lint-staged` cmd and hit interesting behavior. it only works if i provide an extra flag:

```bash
jcs$yarn lint-check
yarn run v1.22.19
$ prettier --check docs/**/*.md --ignore-unknown && prettier --check docs/**/*.mdx --ignore-unknown
Checking formatting...
[warn] docs/developers/bubs-testnet.md
[warn] Code style issues found in the above file. Run Prettier to fix.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
jcs$git add docs/developers/bubs-testnet.md
jcs$npx lint-staged                        
✔ Preparing lint-staged...
✔ Running tasks for staged files...
✖ Prevented an empty git commit!
✔ Reverting to original state because of errors...
✔ Cleaning up temporary files...

  ⚠ lint-staged prevented an empty git commit.
  Use the --allow-empty option to continue, or check your task configuration

jcs$npx lint-staged
✔ Preparing lint-staged...
✔ Running tasks for staged files...
✖ Prevented an empty git commit!
✔ Reverting to original state because of errors...
✔ Cleaning up temporary files...

  ⚠ lint-staged prevented an empty git commit.
  Use the --allow-empty option to continue, or check your task configuration

jcs$npx lint-staged --allow-empty
✔ Preparing lint-staged...
✔ Running tasks for staged files...
✔ Applying modifications from tasks...
✔ Cleaning up temporary files...
```

do you know why?

jcstein reviewed (approved):
LGTM. thanks again for your contribution! much needed improvements to our docs setup.

gulshanpr commented:
> jcs$npx lint-staged
> ✔ Preparing lint-staged...
> ✔ Running tasks for staged files...
> ✖ Prevented an empty git commit!
> ✔ Reverting to original state because of errors...
> ✔ Cleaning up temporary files... 

for testing,
you might be changing the file, that will become as it was before, after formatting (e.g. if `*` is there already and for testing you changed into `-`, and after testing (formatting) with `npx lint-staged` it will become again `*` and it was already there before you made the change) that's why it is giving the error `✖ Prevented an empty git commit!` 

for new changes it will not give the error and work perfectly fine.

P.S. : i was writing this message, why it is producing this error and you merged it 😅

gulshanpr commented:
> LGTM. thanks again for your contribution! much needed improvements to our docs setup.

thank you @jcstein for merging this, it was my first major contribution in open source as well as celestia.

jcstein commented:
can you help me understand if husky is working? i just opened a branch but was able to commit improperly formatted code

github.com/celestiaorg/docs/tree/jcs/test-husky

gulshanpr commented:
>this is poorly formatted  
>text 
>
>with [incomplete][links] and

I'm not sure why this let you commit inspite of having incomplete links, (I'll look into it soon)
but in commit you removed `&& git add . && git commit --no-verify -m 'Apply linting and formatting` as I said this `npx lint-staged` will format the staged file but will not add and commit along with the commit you did without `&& git add . && git commit --no-verify -m 'Apply linting and formatting`. Husky is doing its pre-commit work, if you check `git status ` you will see a modified (formatted) file, but it didn't committed along with the commit

jcstein commented:
> thank you @jcstein for merging this, it was my first major contribution in open source as well as celestia.

you’re welcome. thanks again for working on this!
